crates/turborepo-lib/src/daemon/server.rs

```diff
@@ -0,0 +1,313 @@
+        }
+
+        // here the stop token is dropped, and the pid lock is dropped
+        // causing them to be cleaned up
+    }
+
+    fn bump_timeout(&self) {
+        self.timeout_time.reset(self.timeout);
+    }
+}
+
+#[tonic::async_trait]
+impl<T: Watcher + Send + 'static> proto::turbod_server::Turbod for DaemonServer<T> {
+    async fn hello(
+        &self,
+        request: tonic::Request<proto::HelloRequest>,
+    ) -> Result<tonic::Response<proto::HelloResponse>, tonic::Status> {
+        self.bump_timeout();
+
+        if request.into_inner().version != get_version() {
+            return Err(tonic::Status::unimplemented("version mismatch"));
+        } else {
+            Ok(tonic::Response::new(proto::HelloResponse {}))
+        }
+    }
+
+    async fn shutdown(
+        &self,
+        _request: tonic::Request<proto::ShutdownRequest>,
+    ) -> Result<tonic::Response<proto::ShutdownResponse>, tonic::Status> {
+        self.shutdown.lock().await.take().map(|s| s.send(()));
+
+        // if Some(Ok), then the server is shutting down now
+        // if Some(Err), then the server is already shutting down
+        // if None, then someone has already called shutdown
+        Ok(tonic::Response::new(proto::ShutdownResponse {}))
+    }
+
+    async fn status(
+        &self,
+        _request: tonic::Request<proto::StatusRequest>,
+    ) -> Result<tonic::Response<proto::StatusResponse>, tonic::Status> {
+        self.bump_timeout();
+
+        Ok(tonic::Response::new(proto::StatusResponse {
+            daemon_status: Some(proto::DaemonStatus {
+                uptime_msec: self.start_time.elapsed().as_millis() as u64,
+                log_file: self.log_file.to_str().unwrap().to_string(),
+            }),
+        }))
+    }
+
+    async fn notify_outputs_written(
+        &self,
+        request: tonic::Request<proto::NotifyOutputsWrittenRequest>,
+    ) -> Result<tonic::Response<proto::NotifyOutputsWrittenResponse>, tonic::Status> {
+        self.bump_timeout();
+
+        let inner = request.into_inner();
+        self.watcher
+            .watch_globs(inner.hash, inner.output_globs, inner.output_exclusion_globs)
+            .await;
+
+        Ok(tonic::Response::new(proto::NotifyOutputsWrittenResponse {}))
+    }
+
+    async fn get_changed_outputs(
+        &self,
+        request: tonic::Request<proto::GetChangedOutputsRequest>,
+    ) -> Result<tonic::Response<proto::GetChangedOutputsResponse>, tonic::Status> {
+        self.bump_timeout();
+
+        let inner = request.into_inner();
+        let changed = self
+            .watcher
+            .changed_globs(&inner.hash, HashSet::from_iter(inner.output_globs))
+            .await;
+
+        Ok(tonic::Response::new(proto::GetChangedOutputsResponse {
+            changed_output_globs: changed.into_iter().collect(),
+        }))
+    }
+}
+
+impl<T: Watcher> NamedService for DaemonServer<T> {
+    const NAME: &'static str = "turborepo.Daemon";
+}
+
+#[cfg(test)]
```

Comment:
I notice we dropped the unit tests for signal handling and cleanup on panic. Unsure if we get this functionality for free with tonic, but if not we should probably try to port them just to make sure there aren't any regressions.

Reply:
Yeah mainly because I wasn't sure about the best way to do those. I was hoping to replicate your integration tests but since they are flakey I'm not sure what we should do here. Putting a note to discus later.